File: cse250/StringClient.java
/** File "StringClient.java".  Java version of the "Stack<T>" class in a
    textbook by Ravi Sethi, but with the argument type T fixed as String.
    By KWR for CSE250, Fall 2009, to translate into C++ as-is (Assignment 1)
    CSE250 Fall 2010-->this became a recitation exercise.
    Later we will turn it into a C++ template class Stack<T>.
 */


import java.io.*;   // C++: use #include <iostream>
                    // Also #include <vector> and #include <string>

class StringStack {
   int top; 
   final int maxSize = 101;                 //C++: must init by constructor?
   String elements[] = new String[maxSize]; //C++: use vector<string>*
                                            //Nerdy: vector<const string*>* !
   //C++: ^^^^must initialize in constructor!: new vector<string>(maxSize)


   //CLASS INV: top is the /first free cell/, empty stack is top == 0;


   /** Create empty stack
    */
   public StringStack() {top = 0;}          

   public void push(String c) {
      elements[top++] = c;        //C++: elements->at(top++) = c;  is OK
   }
   public String pop() {
      if (top == 0) {
         System.err.println("Attempt to pop from empty stack."); 
         return "";               //System.err.print ~= cerr << in C++
      } else {
         return elements[--top];  //Is [--top] bad style per text, page 22?
      }                           //Well, as we'll see , it *is* STL-style!
   }                              //But you need the CLASS INV to be sure...
   public int size() { return top; }
}

public class StringClient {               //use just "int main()" in C++
   /** Test-driving code for StringStacks
    */
   public static void main(String args[]) {
      StringStack testStack = new StringStack();
      System.out.print("Pushing \"AAA\" ... ");     //use cout << ... 
      testStack.push("AAA"); 
      System.out.println("Pushing \"BBB\" ... ");   //this tacks on ...<< endl;
      testStack.push("BBB"); 
      System.out.println("Size is now " + testStack.size());
      testStack.pop();               // ^ chain with "<<" instead in C++
      System.out.println("Pop---size is now " + testStack.size());
      String c = testStack.pop();
      System.out.print("I popped the string \"");
      System.out.println(c + "\", size now " + testStack.size());
      System.out.println("Can I pop again?");
      String d = testStack.pop();
      System.out.println("Oops!  I got: \"" + d + "\"");
   }
}

File: cse250/StringClient.cpp
/** File "StringClient.cpp", by KWR for CSE250, Fall 2009.
    Answer Key for Assignment 1, part (4)---with one thematic twist:
    Six lines, all in the StringStack class, are changed from what the
    Java version had in /meaning/, not just syntax.  Can you find them?

    Some alternative translation choices as-noted were also fine, but
    (to save room), the "const string*" option is not shown.
 */

#include <iostream>
#include <vector>       
#include <string>    //not needed on timberlake, *is* needed in Visual C++

using namespace std;

class StringStack {
   int top; 
   //static const int maxSize = 101;        //AOK, but not what Java did
   const int maxSize;                       //initialized by constructor

   vector<string>* elements;

   // CLASS INV': top indexes the /top element/, so empty stack is top == -1;

 public:

   /** Create empty stack
    */
   StringStack() : top(-1), maxSize(101),
                   elements(new vector<string>(maxSize)) { }

   virtual ~StringStack() { cout << "Hi There!" << endl; delete(elements); }   //AOK to omit this line, despite warning
   virtual void push(string c) { elements->at(++top) = c; }
   virtual string pop() {
      if (top == -1) {
         cerr << "Attempt to pop from empty stack." << endl;
         return "";  //System.err.print ~= cerr << in C++
      } else {
         return elements->at(top--); 
      }                             
   }                               
   virtual int size() const { return top+1; }
};

//Extra Credit answer is in StringClientXC.cpp

/** Test-driving code for StringStacks.
 */
int main() {
   StringStack* testStack = new StringStack();
   cout << "Pushing \"AAA\" ... ";        //no "endl", keep next on same line
   testStack->push("AAA"); 
   cout << "Pushing \"BBB\" ... " << endl;
   testStack->push("BBB"); 
   cout << "Size is now " << testStack->size() << endl;
   testStack->pop();
   cout << "Pop---size is now " << testStack->size() << endl;
   string c = testStack->pop();
   cout << "I popped the string \""
        << c << "\", size now " << testStack->size() << endl;
   cout << "Can I pop again?" << endl;
   string d = testStack->pop();
   cout << "Oops!  I got: \"" << d << "\"" << endl;
   //delete(testStack);
   StringStack test2 = (*testStack);
   return (0);
}
   
/*------------------------------------------------------------------
Alternative: one could place a trailing "\n" in the strings, e.g.
   ...
   cout << "Pushing \"BBB\" ... \n";
   ...
   cout << "Size is now " << testStack->size() << "\n";

But in the former, doing \n might add to the confusion of the escaped quotes.
And in the latter, some think tacking a single "\n" like that looks tacky.
Generally, using "endl" is preferred.
*///----------------------------------------------------------------
